services/reddit/reddit-base.js

```diff
@@ -0,0 +1,91 @@
+import Joi from 'joi'
+import { BaseJsonService } from '../index.js'
+
+const tokenSchema = Joi.object({
+  access_token: Joi.string().required(),
+  expires_in: Joi.number(),
+  scope: Joi.string(),
+  token_type: Joi.string(),
+})
+
+// Abstract class for Reddit badges
+// Authorization flow based on https://github.com/reddit-archive/reddit/wiki/OAuth2#application-only-oauth.
+export default class RedditBase extends BaseJsonService {
+  static category = 'social'
+
+  static auth = {
+    userKey: 'reddit_client_id',
+    passKey: 'reddit_client_secret',
+    authorizedOrigins: ['https://www.reddit.com'],
+    isRequired: false,
+  }
+
+  constructor(...args) {
+    super(...args)
+    if (!RedditBase._redditToken && this.authHelper.isConfigured) {
+      RedditBase._redditToken = this._getNewToken()
+    }
+  }
+
+  async _getNewToken() {
+    const tokenRes = await super._requestJson(
+      this.authHelper.withBasicAuth({
+        schema: tokenSchema,
+        url: 'https://www.reddit.com/api/v1/access_token',
+        options: {
+          method: 'POST',
+          body: 'grant_type=client_credentials',
+        },
+        httpErrors: {
+          401: 'invalid token',
+        },
+      }),
+    )
+
+    // replace the token when we are 80% near the expire time
+    // 2147483647 is the max 32-bit value that is accepted by setTimeout(), it's about 24.9 days
+    const replaceTokenMs = Math.min(
+      tokenRes.expires_in * 1000 * 0.8,
+      2147483647,
+    )
+    const timeout = setTimeout(() => {
+      RedditBase._redditToken = this._getNewToken()
+    }, replaceTokenMs)
+
+    // do not block program exit
+    timeout.unref()
```

Comment:
This approach using `setTimeout()` isn't wrong, and I'm not saying we have to change it.
That said, just as a data point: We implemented something quite similar to this in the JWT auth helper.
https://github.com/badges/shields/blob/41d072e1c93e997e8ac5f9c31cb763e351800bdb/core/base-service/auth-helper.js#L256-L318
One thing that is quite nice about that approach is that the control flow is a bit more linear.
Dunno if you reckon that is any better/worse? If you just want to roll with this then I think it is ok.

Reply:
My understanding is that this approach, which is by and large taken from the Twitch implementation, has the advantage of not slowing any badge requests by always ensuring there's a fresh token retrieved asynchronously.

I agree that the JWT code is indeed a bit more linear, but unless I'm mistaken, some badge requests will effectively need to request a fresh token before hitting a second API to retrieve the actual data. Re-reading through the JWT code, I'm also a bit worried that it considers that "the token valid if the expiry datetime is later than (now + 1 minute)"; doesn't that mean that every now and again badges requests will fail if they fall in that 1-minute interval? 🤔